src/codegen.cpp

```diff
@@ -4551,112 +4603,51 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaval)
-
-            size_t nsig = 1 + jl_svec_len(argt_typ->parameters);
-            sig_args = jl_alloc_svec_uninit(nsig);
-            jl_svecset(sig_args, 0, closure_t);
-            for (size_t i = 0; i < jl_svec_len(argt_typ->parameters); ++i) {
-                jl_svecset(sig_args, 1+i, jl_svecref(argt_typ->parameters, i));
-            }
-            li->specTypes = (jl_value_t*)jl_apply_tuple_type_v(jl_svec_data(sig_args), nsig);
-            jl_gc_wb(li, li->specTypes);
-
-            std::tie(closure_m, closure_decls) = emit_function(li, closure_src,
-                ub.constant, ctx.emission_context, jl_unbox_bool(isva.constant));
+            JL_GC_PUSH2(&closure_t, &env_t);
 
             jl_value_t **env_component_ts = (jl_value_t**)alloca(sizeof(jl_value_t*) * (nargs-5));
             for (size_t i = 0; i < nargs - 5; ++i) {
                 env_component_ts[i] = argv[5+i].typ;
             }
 
             env_t = jl_apply_tuple_type_v(env_component_ts, nargs-5);
-            jl_cgval_t env;
-            // TODO: Inline the env at the end of the opaque closure and generate a descriptor for GC
+            // we need to know the full env type to look up the right specialization
             if (jl_is_concrete_type((jl_value_t*)env_t)) {
-                env = emit_new_struct(ctx, (jl_value_t*)env_t, nargs-5, &argv.data()[5]);
-            }
-            else {
-                Value *env_val = emit_jlcall(ctx, jltuple_func, V_rnull,
-                    &argv[5], nargs-5, JLCALL_F_CC);
-                env = mark_julia_type(ctx, env_val, true, env_t);
-            }
-
-            assert(closure_decls.functionObject != "jl_fptr_sparam");
-            bool isspecsig = closure_decls.functionObject != "jl_fptr_args";
+                // TODO: Inline the env at the end of the opaque closure and generate a descriptor for GC
+                jl_cgval_t env = emit_new_struct(ctx, (jl_value_t*)env_t, nargs-5, &argv.data()[5]);
+
+                jl_tupletype_t *argt_typ = (jl_tupletype_t *)argt.constant;
+                closure_t = jl_apply_type2((jl_value_t*)jl_opaque_closure_type, (jl_value_t*)argt_typ, ub.constant);
+
+                jl_method_t *closure_method = (jl_method_t *)source.constant;
+                Function *F = get_oc_function(ctx, closure_method, env_t, argt_typ, ub.constant, jl_unbox_bool(isva.constant));
+                if (F) {
+                    jl_cgval_t jlcall_ptr = mark_julia_type(ctx, F, false, jl_voidpointer_type);
+                    jl_cgval_t world_age = mark_julia_type(ctx,
+                        tbaa_decorate(tbaa_gcframe,
+                                      ctx.builder.CreateAlignedLoad(ctx.world_age_field, Align(sizeof(size_t)))),
+                        false,
+                        jl_long_type);
+                    jl_cgval_t fptr = mark_julia_type(ctx, (llvm::Value*)Constant::getNullValue(T_size), false, jl_voidpointer_type);
```

Comment:
Note I'm just setting the specptr field to NULL here, but AFAICT it's not used yet.

Reply:
I think that's true, but the idea was that even without further type information, calls to concretely-typed opaque closures should use the specptr, and we may want to implement that in short order to avoid having to box everything at the boundry.